### PR TITLE
Add option to generate bitflip visualisations

### PR DESCRIPTION
**Pull request by piotr-binkowski**
This PR adds support for generating visual representation of error lists by adding `--plot` switch when calling `rowhammer.py` or `hw_rowhammer.py`

At the moment two representations are generated:

* 2D Histogram
image

* JSON files used with [Symbiflow DB Visualiser](https://github.com/antmicro/symbiflow-database-visualizer)
image

To view results using DB Visualiser you need to:

* Clone and build the visualiser with
```
git clone https://github.com/antmicro/symbiflow-database-visualizer
cd symbiflow-database-visualizer
npm run build
```
* Run `rowhammer.py` or `hw_rowhammer.py` with `--plot`
* Copy generated JSON files from `vis` directory to `/path/to/symbiflow-database-visualizer/dist/production/`
* Start a simple HTTP server inside the `production` directory: `python -m http.server 8080`

### CODE REVIEW

File: rowhammer_tester/scripts/rowhammer.py [hunk 1]
```diff
@@ -113,7 +113,18 @@ def errors_bitcount(self, row_errors):
...
+    def bitflip_list(val, exp):
+        expr = f'{val ^ exp:#0{len(bin(exp))}b}'
+        return [i for i, c in enumerate(expr[2:]) if c == '1']
+
+    def display_errors(self, row_errors, target_rows=None):
+
+        COLS = 2**self.settings.geom.colbits
+        ROWS = 2**self.settings.geom.rowbits
+
+        mem = [[[] for i in range(COLS)] for j in range(ROWS)]
```
Comment (jedrzejboczar) [inline rowhammer_tester/scripts/rowhammer.py:126]: This takes a quite long time.

File: rowhammer_tester/scripts/rowhammer.py [hunk 2]
```diff
@@ -211,6 +218,156 @@ def payload_executor_attack(self, read_count, row_tuple):
...
+
+    for i in range(rows):
+        if max([len(data[i][j]) for j in range(cols)]) == 0:
+            continue
+        else:
+            last_row = i
+            if first_row is None:
+                first_row = i
+            for j in range(0, cols, step):
+                flips_list = data[i][j:j + step]
```
Comment (jedrzejboczar) [inline rowhammer_tester/scripts/rowhammer.py:300]: This slicing fails with defaultdict approach.

**Review by jedrzejboczar — COMMENTED** [review]
I tested this and everything works well, but I noticed that the list comprehension to create `mem` (rowhammer.py:126) takes a noticeable time (e.g. ~8 seconds), even without `--plot`. It should be possible to avoid it, at least when when the `--plot` argument is not present. I also tried using `collections.defaultdict`:
```python
from collections import defaultdict
mem = defaultdict(lambda: defaultdict(list))
```
because `mem` is usually sparse (only a few errors). Then rows and cols would have to be passed to plot(). This however fails when slicing in _export_vis but I think it should be possible to modify the code to work with it.

**Comment by jedrzejboczar** [comment]
I updated the docs with instructions on how to use the plotting functionality. As now the slowdown happens only when using the `--plot` argument, we can merge this PR. In the future we can try to speed things up, possibly with the defaultdict-based approach.

**Comment by tmichalak** [comment]
Superseded by #121. Closing.

**Pull request closed without merging**